test/http.js

```diff
@@ -110,6 +120,40 @@ test('response contains url', async t => {
 	t.is((await got(s.url)).url, `${s.url}/`);
 });
 
+test('cancel in-progress request', async t => {
+	const p = got(`${s.url}`, {
+		// Readable stream we never close
+		body: new Readable({
+			read() {}
+		}),
+		timeout: 4000,
+		retries: 0
+	});
+
+	setTimeout(() => {
+		p.cancel();
+	}, 100);
+	await t.throws(p, PCancelable.CancelError);
+});
+
+test('cancel immediately', async t => {
+	const p = got(`${s.url}/store-id`);
+	p.cancel();
+	await t.throws(p, PCancelable.CancelError);
+
+	await delay(100);
+	t.false(cancelRequestCompleted);
+});
+
+test.failing('cancel throws after request completed', async t => {
```

Comment:
It works exactly like trying to reject a settled promise. When I promise is settled, it's by the name of it *settled*. Same with cancel. If the user use `.cancel()`, they should be aware they might call it too late and handle that themselves if they need. We can document this well. They can even check the promise whether it was canceled or not with `promise.canceled` (Which is `false` if you try canceling it after the promise is settled). I think `p.cancel()` throwing based on timing could lead to race issues and would be surprising to the user.

Please call me out if you think I'm totally wrong in this. I'm happy to discuss.